src/Microsoft.AspNet.Mvc.Xml/EnumerableWrapperProvider.cs

```diff
@@ -0,0 +1,87 @@
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    public class EnumerableWrapperProvider : IWrapperProvider
+    {
+        private readonly WrapperProviderContext _context;
+        private readonly Type _sourceIEnumerableGenericType;
+        private readonly IEnumerable<IWrapperProviderFactory> _wrapperProviderFactories;
+
+        public EnumerableWrapperProvider(
+            [NotNull] Type sourceIEnumerableGenericType,
+            IEnumerable<IWrapperProviderFactory> wrapperProviderFactories,
+            [NotNull] WrapperProviderContext context)
+        {
+            _sourceIEnumerableGenericType = sourceIEnumerableGenericType;
+            _wrapperProviderFactories = wrapperProviderFactories;
+            _context = context;
+        }
+
+        /// <inheritdoc />
+        public Type WrappingType
+        {
+            get
+            {
+                IWrapperProvider elementWrapperProvider = null;
+                return GetWrappingEnumerableType(out elementWrapperProvider);
+            }
+        }
+
+        /// <inheritdoc />
+        public object Wrap(object original)
+        {
+            if (original != null)
+            {
+                IWrapperProvider elementWrapperProvider = null;
+                var wrappingEnumerableType = GetWrappingEnumerableType(out elementWrapperProvider);
+
+                var wrappingEnumerableTypeConstructor = wrappingEnumerableType.GetConstructor(new[]
+                    {
+                        _sourceIEnumerableGenericType,
+                        typeof(IWrapperProvider)
+                    });
+
+                return wrappingEnumerableTypeConstructor.Invoke(new object[]
+                    {
+                        original,
+                        elementWrapperProvider
+                    });
+            }
+
+            return null;
+        }
+
+        private Type GetWrappingEnumerableType(out IWrapperProvider elementWrapperProvider)
+        {
+            var declaredElementType = _sourceIEnumerableGenericType.GetGenericArguments()[0];
+
+            // Since the T itself could be wrapped, get the wrapping type for it
+            var wrapperProviderContext = new WrapperProviderContext(declaredElementType, _context.IsSerialization);
+
+            var wrappedOrDeclaredElementType = declaredElementType;
+
+            elementWrapperProvider = null;
+            foreach (var wrapperProviderFactory in _wrapperProviderFactories)
+            {
+                elementWrapperProvider = wrapperProviderFactory.GetProvider(wrapperProviderContext);
+                if (elementWrapperProvider != null)
+                {
+                    var wrappingType = elementWrapperProvider.WrappingType;
+                    if (wrappingType != null)
+                    {
+                        wrappedOrDeclaredElementType = wrappingType;
+                    }
+
+                    break;
```

Comment:
move all of this logic into `EnumerableWrapperProviderFactory.GetProvider` and pass it to the constructor of this class